blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
+		calcReachabilityTreeNodeSizes(reindexRootChildNodesBeforeChosen)
+
+	reindexRootStart := reindexRoot.interval.start
+	targetRangeBeforeReindexRootStart := reindexRootStart + reachabilityReindexSlack
+	targetRangeBeforeReindexRootEnd := targetRangeBeforeReindexRootStart + reindexRootChildNodesBeforeChosenSizesSum - 1
+	intervalBeforeReindexRootStart := newReachabilityInterval(targetRangeBeforeReindexRootStart, targetRangeBeforeReindexRootEnd)
+
+	modifiedTreeNodes, err = dag.propagateChildIntervals(intervalBeforeReindexRootStart, reindexRootChildNodesBeforeChosen,
+		reindexRootChildNodesBeforeChosenSizes, reindexRootChildNodesBeforeChosenSubtreeSizeMaps)
+	if err != nil {
+		return 0, nil, err
+	}
+	return reindexRootChildNodesBeforeChosenSizesSum, modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tightenReachabilityTreeIntervalsAfterChosenReindexRootChild(
+	reindexRoot *reachabilityTreeNode, reindexRootChildNodesAfterChosen []*reachabilityTreeNode) (
+	reindexRootChildNodesAfterChosenSizesSum uint64, modifiedTreeNodes map[*reachabilityTreeNode]struct{}, err error) {
+
+	reindexRootChildNodesAfterChosenSizes, reindexRootChildNodesAfterChosenSubtreeSizeMaps, reindexRootChildNodesAfterChosenSizesSum :=
+		calcReachabilityTreeNodeSizes(reindexRootChildNodesAfterChosen)
+
+	reindexRootEnd := reindexRoot.interval.end
+	targetRangeAfterReindexRootEnd := reindexRootEnd - reachabilityReindexSlack
+	targetRangeAfterReindexRootStart := targetRangeAfterReindexRootEnd - reindexRootChildNodesAfterChosenSizesSum
+	intervalAfterReindexRootEnd := newReachabilityInterval(targetRangeAfterReindexRootStart, targetRangeAfterReindexRootEnd-1)
+
+	modifiedTreeNodes, err = dag.propagateChildIntervals(intervalAfterReindexRootEnd, reindexRootChildNodesAfterChosen,
+		reindexRootChildNodesAfterChosenSizes, reindexRootChildNodesAfterChosenSubtreeSizeMaps)
+	if err != nil {
+		return 0, nil, err
+	}
+	return reindexRootChildNodesAfterChosenSizesSum, modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) expandReachabilityTreeIntervalInChosenReindexRootChild(reindexRoot *reachabilityTreeNode,
+	chosenReindexRootChild *reachabilityTreeNode, reindexRootChildNodesBeforeChosenSizesSum uint64,
+	reindexRootChildNodesAfterChosenSizesSum uint64) (map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	reindexRootStart := reindexRoot.interval.start
+	reindexRootEnd := reindexRoot.interval.end
+	targetRangeForReindexRootChildStart :=
+		reindexRootStart + reindexRootChildNodesBeforeChosenSizesSum + reachabilityReindexSlack
+	targetRangeForReindexRootChildEnd :=
+		reindexRootEnd - reindexRootChildNodesAfterChosenSizesSum - reachabilityReindexSlack - 1
+	newReindexRootChildInterval := newReachabilityInterval(targetRangeForReindexRootChildStart, targetRangeForReindexRootChildEnd)
+
+	if targetRangeForReindexRootChildStart > reindexRootStart || targetRangeForReindexRootChildEnd < reindexRootEnd {
```

Comment:
**CRITICAL ERROR**:  the comparison here should be with `prevReindexRootChildStart` and `prevReindexRootChildEnd`
which should be set `prevReindexRootChildStart := chosenReindexRootChild.interval.start` and `prevReindexRootChildEnd := chosenReindexRootChild.interval.end`